### PR TITLE
FE-1757 - Enhance the coding challenges we have available for interviewing candidates

### PR DESCRIPTION
**Pull request by nstanard**
## What Changed

 - Finished the solution
 - Added test coverage for the finished solution
 - Updated the notes/instructions

### CODE REVIEW

File: src/components/App.jsx [hunk 1]
```diff
@@ -1,5 +1,27 @@
 import React from "react";
 import AppLayout from "./AppLayout";
+// https://design.sparkpost.com/components
+// import { Badge, Box, CodeBlock, Stack, Text } from '@sparkpost/matchbox';
+
+// Instructions:
+// Step 1. Get the data and finish/display the Navigation (./mocks/nav.png)
```
Comment (nstanard) [inline src/components/App.jsx:7]: Should I make the same comment here that the readme makes - about `GET /data/eventsData.json` ?

File: src/components/App.jsx [hunk 2]
```diff
@@ -1,5 +1,27 @@
 import React from "react";
 import AppLayout from "./AppLayout";
+// https://design.sparkpost.com/components
+// import { Badge, Box, CodeBlock, Stack, Text } from '@sparkpost/matchbox';
+
+// Instructions:
+// Step 1. Get the data and finish/display the Navigation (./mocks/nav.png) 
+// Step 2. Clicking the navigation displays the related details (./mocks/navAndDetails.png)
```
Comment (llacava) [inline src/components/App.jsx:8]: Maybe `// Step 2. Clicking the navigation displays the related details in the main view(./mocks/navAndDetails.png)` ? Or even `// Step 2. Clicking the navigation displays the related details in the left panel(./mocks/navAndDetails.png)`

File: src/components/App.jsx (hunk 1 above)
Comment (llacava) [reply src/components/App.jsx:7]: Yeah! I think so

File: src/components/Main.jsx [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export default function() {
+    // TODO: Step 2 - clicking nav items shows the details
+    return null;
+}
+
+// Open Question: Should we do this for the candidate or leave this for them to create?
```
Comment (llacava) [inline src/components/Main.jsx:6]: I think I'd be better to let it free for the candidate, maybe he/she will do it all in the same file or think in another way of componentization

File: src/components/Nav.jsx [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import NavItem from './NavItem';
+
+export default function() {
+    // TODO: Step 1 - Display the nav items after getting the data
+    return null;
+}
+
+// Open Question: Should we do this for the candidate or leave this for them to create?
```
Comment (llacava) [inline src/components/Nav.jsx:8]: Same thing from above :D

**Review by llacava — APPROVED** [review]

File: src/components/NavItem.jsx [hunk 5]
```diff
@@ -40,4 +41,4 @@ NavItem.propTypes = {
   selected: PropTypes.bool,
 };
 
-export default NavItem;
+export default React.memo(NavItem);
```
Comment (nstanard) [inline src/components/NavItem.jsx:44]: talking point: arrow functions with React.memo

File: src/components/NavItem.jsx [hunk 6]
```diff
@@ -21,8 +21,9 @@ const NavItem = ({ children, onClick, selected }) => {
     onClick();
   };
 
+  // Talking point: li nav items
   return (
-    <Box marginBottom="200">
+    <Box marginBottom="200" as="li">
```
Comment (nstanard) [inline src/components/NavItem.jsx:26]: Talking point: li for the nav items (expect <nav> symantic markup with <ul> parent of nav items)

File: src/components/App.jsx [hunk 7]
```diff
@@ -1,11 +1,73 @@
...
+  const handleOptionChange = useCallback((e) => {
+    setGroup(e.target.value);
+  }, []);
+
+  const handleNavItemSelect = useCallback((key, label) => {
+    setDisplayContent(rawData[key]);
+  }, [rawData]);
+
+  useEffect(async () => {
+    const data = await fetch('/data/eventTypes.json').then(res => res.json());
```
[the remark this answers is not in the document]
Comment (llacava) [reply src/components/App.jsx:48]: Uhmmm, maybe we can create a services folder and place API calls there?

File: src/components/Nav.jsx [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { Box, Select } from '@sparkpost/matchbox';
+import NavItem from './NavItem';
+
+export default function({ handleOptionChange, groupOptions, group: navGroup, navItems, handleSelect }) { 
+    if (!navItems) return null;
+    // Talking point: key in mapping, on click arrow function vs useCallback (for React.memo)
+    return <Box as="nav">
+        {groupOptions && <Select id="nav" options={groupOptions} onChange={handleOptionChange} mb="400"></Select>}
```
Comment (llacava) [inline src/components/Nav.jsx:8]: Hey! I think `Select` is a self closing component, we can change it this line to: 
`{groupOptions && <Select id="nav" options={groupOptions} onChange={handleOptionChange} mb="400"/>} `

File: src/components/tests/App.test.js [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    it("Displays the content on navigation click.", async () => {
+        let wrapper;
+
+        await act(async () => {
+            wrapper = render(<App />);
+        });
+
+        const navigation = screen.getByRole('navigation');
+        const main = screen.getByRole('main');
+        const badge = main.querySelector('h1 + div');
```
Comment (nstanard) [inline src/components/tests/App.test.js:99]: we need to explore getting the badge element another way

File: src/components/App.jsx [hunk 10]
```diff
@@ -1,11 +1,99 @@
...
 const App = () => {
+  const [rawData, setRawData] = useState();
+  const [navItems, setNavItems] = useState();
+  const [displayContent, setDisplayContent] = useState();
+  const [group, setGroup] = useState();
+  const [groupOptions, setGroupOptions] = useState();
+
+  const handleOptionChange = useCallback((e) => {
+    setGroup(e.target.value);
+    // TODO: reset content to first in group
```
Comment (nstanard) [inline src/components/App.jsx:40]: We might want to add one more feature - on select change the main content gets reset to the first nav item. We would add a jest test for that too if we wanted. Or we could intentionally leave this off and ask the user to talk about UI/UX improvements.

**Comment by nstanard** [comment]
@llacava Ready for final review! 🙏

File: src/components/Main.jsx [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import { Badge, Box, CodeBlock, Text } from '@sparkpost/matchbox';
+
+export default function({ content }) { 
+    if (!content) return null;
+    return <Box as="main" padding="400">
+        <Box as="h1">{content.label}</Box>
+        {/* Matchbox is stripping data-testid off of this.... hmmm */}
```
Comment (nstanard) [inline src/components/Main.jsx:7]: @jonambas Please keep me posted if you get this merged/deployed/fixed or w/e needs done. 🙏

**Comment by llacava** [comment]
LGTM! ✅

**Pull request merged**